### PR TITLE
Import the Panfrost kernel driver for FreeBSD.

### PR DESCRIPTION
**Pull request by bukinr**
Tested on rk3399 with wayland (sway compositor), chromium, firefox and mesa-demos.

### CODE REVIEW

File: panfrost/panfrost_features.h [hunk 1]
```diff
@@ -0,0 +1,304 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
Comment (evadot) [inline panfrost/panfrost_features.h:1]: We can't add a GPL-2.0 file here.

File: panfrost/panfrost_issues.h [hunk 2]
```diff
@@ -0,0 +1,253 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
Comment (evadot) [inline panfrost/panfrost_issues.h:1]: Same here.

File: panfrost/panfrost_regs.h [hunk 3]
```diff
@@ -0,0 +1,321 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
Comment (evadot) [inline panfrost/panfrost_regs.h:1]: And same here.

File: panfrost/panfrost_features.h (hunk 1 above)
Comment (bukinr) [reply panfrost/panfrost_features.h]: replaced with BSD header

File: panfrost/panfrost_issues.h (hunk 2 above)
Comment (bukinr) [reply panfrost/panfrost_issues.h]: replaced with BSD header

File: panfrost/panfrost_regs.h (hunk 3 above)
Comment (bukinr) [reply panfrost/panfrost_regs.h]: replaced with BSD header

**Comment by jsm222** [comment]
With this latest non GPL version I get CPU 3:  0.0% user,  0.0% nice,  100% system,  0.0% interrupt,  0.0% idle
 an an unusable system. It happens when starting Firefox under xfce4 on Pinebook Pro.. and also whens starting sway. The inital PR with the gpl header files works.  My current-14 is at c9e023541aefdabf0b602e1fd62f5e21a6f8ffa2 
CPU 0:  0.0% user,  0.0% nice,  100% system,  0.0% interrupt,  0.0% idle
  1297 root          3  80    0   270M   141M CPU0     0   2:47 100.18% Xorg. I have tested a few times so either my testing is wrong, or a functional change has sneaked in or it simply does not work with current  at commit c9e023541aefdabf0b602e1fd62f5e21a6f8ffa2. Thanks.

**Comment by jsm222** [comment]
Yes, with that single change it works as well as the earlier versions.

**Comment by bukinr** [comment]
What is your mesa-dri version? I updated locally to latest version (the version used in ports/pkgs) and noticed that new mesa imports panfrost DRM objects into rk3399 DRM, while the rk3399 drm could not do non-contiguous memory because it has no IOMMU implemented in drivers. So if you update mesa and try sway you may see issues in picture

**Comment by bukinr** [comment]
Sorry it is not a mesa3d issue, since I can run panfrost normally on X even with non-contiguous backed pages.
I guess it is the sway issue.

**Comment by bukinr** [comment]
@jsm222 I added usage of vm_page_reclaim_contig() to reclaim pages when you have no free memory.
Please check latest version without modifications.

**Comment by bukinr** [comment]
How much memory do you have in your machine?

**Comment by jsm222** [comment]
4GB I still  have 2096M free when crashing. I will try on the rockpro64 to se if it crashes the same way.

**Comment by jsm222** [comment]
the rockpro64 crashes the same way (same sd card, new u-boot)
Fatal data abort:
  x0: ffffa00006ad9cf0
  x1:                0
  x2:                2
  x3: ffff000040239f00 (ucom_cons_softc + 3ee2cfd0)
  x4:               10
  x5:                0
  x6:              120
  x7:                c
  x8:             1000
  x9: ffff0000e6cc5008 (__stop_set_modmetadata_set + 655250)
 x10:             3fff
 x11:         ffffefff
 x12:                0
 x13:                0
 x14:                0
 x15:               38
 x16: ffff0000e662a3c8 (__stop_set_modmetadata_set + 440)
 x17: ffff0000009525ac (drm_prime_pages_to_sg + 0)
 x18: ffff0000e63af320 (ucom_cons_softc + e4fa23f0)
 x19: ffffa00006ad9cf0
 x20:             4000
 x21: ffff0000e6cc5000 (__stop_set_modmetadata_set + 655248)
 x22:                0
 x23:                1
 x24:              200
 x25:                1
 x26: ffff0000e6603f81 ($d.2 + 0)
 x27: ffffa000c1a6ec00
 x28: ffffa00006e0c008
 x29: ffff0000e63af320 (ucom_cons_softc + e4fa23f0)
  sp: ffff0000e63af320
  lr: ffff0000009525e4 (drm_prime_pages_to_sg + 38)
 elr: ffff000000952634 (drm_prime_pages_to_sg + 88)

**Comment by jsm222** [comment]
> How much memory do you have in your machine?

I get allocation error when npages are 16384, i guees I do not have enough free continuous free memory.

**Comment by evadot** [comment]
Merged in 52e081619446bc895cd9903eeffd551c329d61f5

**Pull request closed without merging**